File: pkg/scanner/logic.go
package scanner

import (
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/zanicool/apex-cli/pkg/crawler"
	"github.com/zanicool/apex-cli/pkg/engine"
	"github.com/zanicool/apex-cli/pkg/oob"
)

// --- Race Condition ---

func scanRaceCondition(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	raceKeywords := []string{"coupon", "redeem", "transfer", "vote", "like", "follow",
		"apply", "claim", "register", "withdraw", "purchase", "checkout",
		"add", "create", "submit", "confirm", "verify", "activate"}

	// Test forms
	for _, form := range crawl.Forms {
		action := form.Action
		isRaceTarget := false
		for _, kw := range raceKeywords {
			if strings.Contains(strings.ToLower(action), kw) {
				isRaceTarget = true
				break
			}
		}
		if !isRaceTarget {
			continue
		}

		data := ""
		for _, inp := range form.Inputs {
			if data != "" {
				data += "&"
			}
			val := inp.Value
			if val == "" {
				val = "test"
			}
			data += inp.Name + "=" + val
		}

		result := raceTest(http, "POST", action, "application/x-www-form-urlencoded", data)
		if result != nil {
			findings = append(findings, *result)
		}
	}

	// Test API endpoints from crawl
	for _, page := range crawl.Pages {
		urlLower := strings.ToLower(page.URL)
		for _, kw := range raceKeywords {
			if strings.Contains(urlLower, kw) && strings.Contains(urlLower, "/api") {
				result := raceTest(http, "POST", page.URL, "application/json", "{}")
				if result != nil {
					findings = append(findings, *result)
				}
				break
			}
		}
	}

	return findings
}

func raceTest(http *engine.HTTPClient, method, url, contentType, body string) *Finding {
	// Fire 20 parallel requests simultaneously
	var wg sync.WaitGroup
	var mu sync.Mutex
	successes := 0
	responses := make([]int, 0, 20)

	for i := 0; i < 20; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			resp := http.Post(url, contentType, body)
			if resp.Err == nil {
				mu.Lock()
				responses = append(responses, resp.StatusCode)
				if resp.StatusCode >= 200 && resp.StatusCode < 400 {
					successes++
				}
				mu.Unlock()
			}
		}()
	}
	wg.Wait()

	// If multiple succeeded AND we got consistent 200s, it's likely a race condition
	if successes > 1 {
		return &Finding{
			Type: "Race Condition — Limit Bypass", Severity: "high",
			URL: url, Detail: fmt.Sprintf("%d/20 parallel requests succeeded — duplicate action possible", successes),
			Template: "apex-race-condition",
		}
	}
	return nil
}

// --- Price Manipulation ---

func scanPriceManipulation(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	priceParams := []string{"price", "amount", "total", "cost", "value", "quantity", "qty", "discount"}

	for u, params := range crawl.Params {
		for _, p := range params {
			pLower := strings.ToLower(p)
			isPriceParam := false
			for _, pp := range priceParams {
				if strings.Contains(pLower, pp) {
					isPriceParam = true
					break
				}
			}
			if !isPriceParam {
				continue
			}

			// Test negative values
			for _, val := range []string{"-1", "0", "0.01", "99999999"} {
				testURL := injectParam(u, p, val)
				resp := http.Get(testURL)
				if resp.Err == nil && resp.StatusCode == 200 &&
					!strings.Contains(strings.ToLower(resp.Body), "invalid") &&
					!strings.Contains(strings.ToLower(resp.Body), "error") {
					findings = append(findings, Finding{
						Type: "Price/Amount Manipulation", Severity: "high",
						URL: testURL, Param: p, Payload: val,
						Detail:   fmt.Sprintf("Server accepted %s=%s without validation", p, val),
						Template: "apex-price-manipulation",
					})
					break
				}
			}
		}
	}
	return findings
}

// --- Payment Flow Bypass ---

func scanPaymentBypass(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	paymentPaths := []string{"/checkout/complete", "/order/confirm", "/payment/success",
		"/api/order/complete", "/api/payment/verify", "/purchase/finalize"}

	for _, page := range crawl.Pages[:min(5, len(crawl.Pages))] {
		baseURL := strings.Split(page.URL, "/")[0] + "//" + strings.Split(page.URL, "/")[2]
		for _, path := range paymentPaths {
			url := baseURL + path
			// Try accessing payment success directly
			resp := http.Get(url)
			if resp.Err == nil && resp.StatusCode == 200 &&
				(strings.Contains(resp.Body, "success") || strings.Contains(resp.Body, "confirmed") ||
					strings.Contains(resp.Body, "order")) {
				findings = append(findings, Finding{
					Type: "Payment Flow Bypass — Direct Access", Severity: "critical",
					URL: url, Detail: "Payment confirmation endpoint accessible without completing payment",
					Template: "apex-payment-bypass",
				})
			}
			// Try POST with manipulated status
			resp2 := http.Post(url, "application/json", `{"status":"paid","amount":0}`)
			if resp2.Err == nil && resp2.StatusCode == 200 {
				findings = append(findings, Finding{
					Type: "Payment Flow Bypass — Status Manipulation", Severity: "critical",
					URL: url, Payload: `{"status":"paid","amount":0}`,
					Template: "apex-payment-status",
				})
			}
		}
	}
	return findings
}

// --- Mass Assignment ---

func scanMassAssignment(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	dangerousFields := []string{"role", "admin", "is_admin", "isAdmin", "privilege",
		"permissions", "user_type", "verified", "is_verified", "balance", "credits", "plan"}

	for _, form := range crawl.Forms {
		action := form.Action
		if !strings.Contains(strings.ToLower(action), "register") &&
			!strings.Contains(strings.ToLower(action), "signup") &&
			!strings.Contains(strings.ToLower(action), "profile") &&
			!strings.Contains(strings.ToLower(action), "update") &&
			!strings.Contains(strings.ToLower(action), "user") {
			continue
		}

		existingFields := make(map[string]bool)
		baseData := make(map[string]string)
		for _, inp := range form.Inputs {
			existingFields[inp.Name] = true
			val := inp.Value
			if val == "" {
				val = "test"
			}
			baseData[inp.Name] = val
		}

		for _, field := range dangerousFields {
			if existingFields[field] {
				continue
			}
			testData := make(map[string]string)
			for k, v := range baseData {
				testData[k] = v
			}
			testData[field] = "true"

			// Build JSON body
			parts := []string{}
			for k, v := range testData {
				parts = append(parts, fmt.Sprintf(`"%s":"%s"`, k, v))
			}
			body := "{" + strings.Join(parts, ",") + "}"

			resp := http.Post(action, "application/json", body)
			if resp.Err == nil && resp.StatusCode >= 200 && resp.StatusCode < 300 &&
				strings.Contains(resp.Body, field) {
				findings = append(findings, Finding{
					Type: "Mass Assignment — " + field, Severity: "high",
					URL: action, Param: field, Payload: field + "=true",
					Detail:   fmt.Sprintf("Server accepted undocumented field '%s'. Privilege escalation possible.", field),
					Template: "apex-mass-assignment",
				})
				break
			}
		}
	}
	return findings
}

// --- Forced Browsing ---

func scanForcedBrowsing(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	var mu sync.Mutex
	sem := make(chan struct{}, cfg.Threads)
	var wg sync.WaitGroup

	adminPaths := []string{
		"/admin", "/admin/", "/administrator", "/dashboard", "/panel",
		"/manage", "/management", "/internal", "/debug", "/console",
		"/api/admin", "/api/internal", "/api/debug", "/graphql",
		"/_debug", "/actuator", "/actuator/env", "/actuator/health",
		"/swagger-ui.html", "/api-docs", "/swagger.json",
		"/.env", "/config", "/phpinfo.php", "/server-status",
		"/wp-admin", "/wp-json/wp/v2/users", "/.git/config",
		"/api/v1/admin", "/api/users", "/metrics", "/health",
	}

	for _, page := range crawl.Pages[:min(5, len(crawl.Pages))] {
		baseURL := strings.Split(page.URL, "/")[0] + "//" + strings.Split(page.URL, "/")[2]
		for _, path := range adminPaths {
			wg.Add(1)
			go func(url string) {
				defer wg.Done()
				sem <- struct{}{}
				defer func() { <-sem }()

				resp := http.Get(url)
				if resp.Err != nil || resp.StatusCode == 404 || resp.StatusCode == 403 || resp.StatusCode == 401 {
					return
				}
				if engine.IsWAFChallenge(resp) {
					return
				}
				// Reject redirects to login/auth pages
				if resp.StatusCode >= 300 && resp.StatusCode < 400 {
					return
				}
				if resp.StatusCode == 200 && len(resp.Body) > 100 {
					// Reject if it's actually a login/auth page
					bodyLower := strings.ToLower(resp.Body)
					if strings.Contains(bodyLower, "sign in") || strings.Contains(bodyLower, "log in") || strings.Contains(bodyLower, "login") {
						return
					}
					// Verify it's not a generic page
					if strings.Contains(strings.ToLower(resp.Body), "admin") ||
						strings.Contains(strings.ToLower(resp.Body), "dashboard") ||
						strings.Contains(strings.ToLower(resp.Body), "actuator") ||
						strings.Contains(resp.Body, "SECRET") ||
						strings.Contains(resp.Body, "password") {
						mu.Lock()
						findings = append(findings, Finding{
							Type: "Forced Browsing — Sensitive Endpoint", Severity: "high",
							URL: url, Detail: fmt.Sprintf("Accessible without auth (status %d, %d bytes)", resp.StatusCode, resp.Size),
							Template: "apex-forced-browsing",
						})
						mu.Unlock()
					}
				}
			}(baseURL + path)
		}
	}
	wg.Wait()
	return findings
}

// --- IDOR via UUID Prediction ---

func scanIDORUUID(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, _ *oob.Client) []Finding {
	var findings []Finding
	// Already covered in scanner.go scanIDOR, this adds UUID-specific logic
	// Look for sequential UUIDs (v1 UUIDs are time-based and predictable)
	for u := range crawl.Params {
		if strings.Contains(u, "-") {
			parts := strings.Split(u, "/")
			for _, part := range parts {
				if len(part) == 36 && strings.Count(part, "-") == 4 {
					// UUID v1 check (first 8 chars are time-based)
					if part[14] == '1' { // Version 1
						findings = append(findings, Finding{
							Type: "IDOR Risk — UUID v1 (Predictable)", Severity: "medium",
							URL: u, Detail: "UUID v1 detected — time-based and predictable. Adjacent UUIDs can be guessed.",
							Template: "apex-idor-uuid-v1",
						})
					}
					break
				}
			}
		}
	}
	return findings
}

// Placeholder for timing
var _ = time.Second

File: src/scanners/logic.cpp
/// @file scanners/logic.cpp
/// @brief Business logic flaws with differential, endpoint-specific evidence.
#include <future>
#include <thread>

#include "scanner_base.hpp"
#include "../response_validator.hpp"

namespace apex {
namespace {

std::string lower_body(const Response& response) {
  std::string value = response.body;
  std::transform(value.begin(), value.end(), value.begin(), ::tolower);
  return value;
}

bool distinct_success(const Response& response, const Response& soft404, int threshold = 80) {
  return response.status_code >= 200 && response.status_code < 300 &&
         (response.status_code != soft404.status_code || responses_differ(response, soft404, threshold));
}

std::vector<Finding> scan_race_condition(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  const std::string base = base_url_from(crawl.urls[0]);
  const std::string probe_path = "/api/apex-race-probe-7f3c9d";
  const auto get_probe = http.get(base + probe_path);
  const auto post_probe = http.post(base + probe_path, R"({"amount":1})", "application/json");
  const std::vector<std::string> paths = {"/api/transfer", "/api/redeem", "/api/coupon", "/api/withdraw"};

  for (const auto& path : paths) {
    const auto discovery = http.get(base + path);
    if (discovery.status_code == 0 || discovery.status_code == 404 ||
        (discovery.status_code == get_probe.status_code && !responses_differ(discovery, get_probe, 50))) continue;

    std::vector<std::future<Response>> futures;
    for (int i = 0; i < 5; ++i) {
      futures.push_back(std::async(std::launch::async, [&http, base, path]() {
        return http.post(base + path, R"({"amount":1})", "application/json");
      }));
    }
    int confirmed_success = 0;
    std::string evidence;
    for (auto& future : futures) {
      const auto response = future.get();
      const std::string body = lower_body(response);
      const bool success_semantics = body.find("success") != std::string::npos || body.find("completed") != std::string::npos ||
                                     body.find("transaction") != std::string::npos || body.find("redeemed") != std::string::npos ||
                                     body.find("transfer_id") != std::string::npos || body.find("order_id") != std::string::npos;
      if (distinct_success(response, post_probe) && success_semantics) {
        ++confirmed_success;
        if (evidence.empty()) evidence = response.body.substr(0, 200);
      }
    }
    if (confirmed_success > 1) {
      findings.push_back({"Race Condition", "high", base + path,
                          "Multiple concurrent requests produced endpoint-specific success responses (" +
                              std::to_string(confirmed_success) + "/5)",
                          "", "", evidence});
    }
  }
  return findings;
}

std::vector<Finding> scan_price_manipulation(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  for (const auto& url : crawl.urls) {
    for (const auto& parameter : crawl.params) {
      if (parameter.url != url) continue;
      const std::string name = parameter.name;
      if (name.find("price") == std::string::npos && name.find("amount") == std::string::npos &&
          name.find("cost") == std::string::npos && name.find("total") == std::string::npos) continue;
      const auto baseline = http.get(parameter.url + "?" + name + "=19.99");
      const auto response = http.get(parameter.url + "?" + name + "=0.01");
      const std::string body = lower_body(response);
      const bool accepted = body.find("0.01") != std::string::npos &&
                            (body.find("total") != std::string::npos || body.find("price") != std::string::npos ||
                             body.find("amount") != std::string::npos || body.find("accepted") != std::string::npos);
      if (response.status_code == 200 && baseline.status_code == 200 && accepted && responses_differ(response, baseline, 10)) {
        findings.push_back({"Price Manipulation", "high", url,
                            "Modified price was reflected in transaction-specific response data", name, "0.01",
                            response.body.substr(0, 200)});
      }
    }
  }
  return findings;
}

std::vector<Finding> scan_mass_assignment(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  const std::string base = base_url_from(crawl.urls[0]);
  const std::vector<std::string> paths = {"/api/user", "/api/profile", "/api/account", "/api/settings"};
  const std::string normal = R"({"name":"test","role":"user","is_admin":false})";
  const std::string echo_marker = "apex_unknown_assignment_control_7f3c9d";
  const std::string escalation =
      R"({"name":"test","role":"admin","is_admin":true,"verified":true,"apex_control":"apex_unknown_assignment_control_7f3c9d"})";
  for (const auto& path : paths) {
    const auto baseline = http.post(base + path, normal, "application/json");
    if (baseline.status_code == 404 || baseline.status_code == 405) continue;
    const auto response = http.post(base + path, escalation, "application/json");
    const bool escalated = response.body.find("\"role\":\"admin\"") != std::string::npos ||
                           response.body.find("\"is_admin\":true") != std::string::npos;
    const bool baseline_admin = baseline.body.find("\"role\":\"admin\"") != std::string::npos ||
                                baseline.body.find("\"is_admin\":true") != std::string::npos;
    if (response.status_code == 200 && escalated && !baseline_admin &&
        response.body.find(echo_marker) == std::string::npos && responses_differ(response, baseline, 10)) {
      findings.push_back({"Mass Assignment", "critical", base + path,
                          "Privilege fields were applied only in the escalated request", "", escalation,
                          "Baseline lacks admin role; escalated response contains applied admin role"});
    }
  }
  return findings;
}

std::vector<Finding> scan_forced_browsing(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  const std::string base = base_url_from(crawl.urls[0]);
  const auto soft404 = http.get(base + "/apex-admin-probe-7f3c9d");
  const std::vector<std::string> paths = {"/admin", "/admin/", "/debug", "/debug/", "/console", "/phpmyadmin",
                                          "/adminer", "/wp-admin", "/manager", "/actuator/env", "/elmah.axd", "/_profiler"};
  for (const auto& path : paths) {
    const auto response = http.get(base + path);
    const std::string body = lower_body(response);
    const bool admin_content = body.find("admin dashboard") != std::string::npos || body.find("phpmyadmin") != std::string::npos ||
                               body.find("adminer") != std::string::npos || body.find("propertysources") != std::string::npos ||
                               body.find("debug toolbar") != std::string::npos || body.find("server information") != std::string::npos;
    const bool login_page = body.find("password") != std::string::npos || body.find("sign in") != std::string::npos ||
                            body.find("log in") != std::string::npos;
    if (distinct_success(response, soft404, 100) && response.body.size() > 100 && admin_content && !login_page) {
      findings.push_back({"Forced Browsing", "high", base + path,
                          "Distinct administrative/debug content accessible without an authentication challenge", "", "",
                          response.body.substr(0, 200)});
    }
  }
  return findings;
}

std::vector<Finding> scan_idor_uuid(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  const std::regex uuid_re(R"([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})", std::regex::icase);
  const std::vector<std::string> private_markers = {"\"email\"", "\"phone\"", "\"address\"", "\"account\"", "\"order\"", "\"billing\""};
  for (const auto& url : crawl.urls) {
    std::smatch match;
    if (!std::regex_search(url, match, uuid_re)) continue;
    const auto baseline = http.get(url);
    std::string modified = url;
    modified.replace(match.position(), match.length(), "00000000-0000-0000-0000-000000000001");
    const auto response = http.get(modified);
    if (response.status_code == 200 && baseline.status_code == 200 && responses_differ(response, baseline, 20) &&
        contains_any(response.body, private_markers)) {
      findings.push_back({"IDOR (UUID)", "high", url,
                          "A different UUID returned distinct private resource data", "", "00000000-0000-0000-0000-000000000001",
                          response.body.substr(0, 200)});
    }
  }
  return findings;
}

std::vector<Finding> scan_payment_bypass(const Config&, HttpClient& http, const CrawlResult& crawl) {
  std::vector<Finding> findings;
  if (crawl.urls.empty()) return findings;
  const std::string base = base_url_from(crawl.urls[0]);
  const auto soft404 = http.get(base + "/apex-payment-complete-probe-7f3c9d");
  const std::vector<std::string> paths = {"/order/confirm", "/checkout/success", "/payment/complete", "/api/order/complete", "/thank-you"};
  for (const auto& path : paths) {
    const auto response = http.get(base + path);
    const std::string body = lower_body(response);
    const bool transaction_data = body.find("order_id") != std::string::npos || body.find("order number") != std::string::npos ||
                                  body.find("payment_status") != std::string::npos || body.find("transaction id") != std::string::npos ||
                                  body.find("receipt") != std::string::npos;
    const bool completion = body.find("paid") != std::string::npos || body.find("payment complete") != std::string::npos ||
                            body.find("order confirmed") != std::string::npos || body.find("thank you for your order") != std::string::npos;
    if (distinct_success(response, soft404, 100) && transaction_data && completion) {
      findings.push_back({"Payment Bypass", "high", base + path,
                          "Post-payment route exposed transaction-specific completion data without a verified payment flow", "", "",
                          response.body.substr(0, 250)});
    }
  }
  return findings;
}

}  // namespace

std::vector<Scanner> register_logic_scanners() {
  return {{"Race Condition", scan_race_condition}, {"Price Manipulation", scan_price_manipulation},
          {"Mass Assignment", scan_mass_assignment}, {"Forced Browsing", scan_forced_browsing},
          {"IDOR (UUID)", scan_idor_uuid}, {"Payment Bypass", scan_payment_bypass}};
}

}  // namespace apex
